### ISSUE TITLE
Upgrade to SpringFox 2.9.0 causes failures in Spring Boot 2.0.2.RELEASE

### ISSUE THREAD

**Issue by james-bryant** [springfox/springfox#2438]
After upgrading to SpringFox 2.9.0, we start getting this exception when we try to start Spring Boot:

```stack trace
org.springframework.beans.factory.BeanDefinitionStoreException: Failed to read candidate component class: file [D:\workspace-git2\cagateway\cagw-web\build\classes\java\main\com\entrust\adminservices\cagw\Swagger2SpringBoot.class]; nested exception is java.lang.IllegalStateException: Error processing condition on com.entrust.adminservices.cagw.Swagger2SpringBoot
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.scanCandidateComponents(ClassPathScanningCandidateComponentProvider.java:454)
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.findCandidateComponents(ClassPathScanningCandidateComponentProvider.java:316)
	at org.springframework.boot.test.context.SpringBootConfigurationFinder.scanPackage(SpringBootConfigurationFinder.java:67)
	at org.springframework.boot.test.context.SpringBootConfigurationFinder.findFromPackage(SpringBootConfigurationFinder.java:59)
	at org.springframework.boot.test.context.SpringBootConfigurationFinder.findFromClass(SpringBootConfigurationFinder.java:52)
	at org.springframework.boot.test.context.SpringBootTestContextBootstrapper.getOrFindConfigurationClasses(SpringBootTestContextBootstrapper.java:239)
	at org.springframework.boot.test.context.SpringBootTestContextBootstrapper.processMergedContextConfiguration(SpringBootTestContextBootstrapper.java:153)
	at org.springframework.test.context.support.AbstractTestContextBootstrapper.buildMergedContextConfiguration(AbstractTestContextBootstrapper.java:395)
	at org.springframework.test.context.support.AbstractTestContextBootstrapper.buildDefaultMergedContextConfiguration(AbstractTestContextBootstrapper.java:312)
	at org.springframework.test.context.support.AbstractTestContextBootstrapper.buildMergedContextConfiguration(AbstractTestContextBootstrapper.java:265)
	at org.springframework.test.context.support.AbstractTestContextBootstrapper.buildTestContext(AbstractTestContextBootstrapper.java:108)
	at org.springframework.boot.test.context.SpringBootTestContextBootstrapper.buildTestContext(SpringBootTestContextBootstrapper.java:97)
	at org.springframework.test.context.TestContextManager.<init>(TestContextManager.java:139)
	at org.springframework.test.context.TestContextManager.<init>(TestContextManager.java:124)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTestContextManager(SpringJUnit4ClassRunner.java:151)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.<init>(SpringJUnit4ClassRunner.java:142)
	at org.springframework.test.context.junit4.SpringRunner.<init>(SpringRunner.java:49)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:488)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:33)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:88)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:59)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:39)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy1.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:146)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:128)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.IllegalStateException: Error processing condition on com.entrust.adminservices.cagw.Swagger2SpringBoot
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:64)
	at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:108)
	at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:88)
	at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:71)
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.isConditionMatch(ClassPathScanningCandidateComponentProvider.java:515)
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.isCandidateComponent(ClassPathScanningCandidateComponentProvider.java:498)
	at org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider.scanCandidateComponents(ClassPathScanningCandidateComponentProvider.java:431)
	... 56 more
Caused by: java.lang.IllegalStateException: No ConfigurableListableBeanFactory available
	at org.springframework.util.Assert.state(Assert.java:73)
	at org.springframework.context.annotation.ConditionEvaluator$ConditionContextImpl.getBeanFactory(ConditionEvaluator.java:206)
	at org.springframework.boot.autoconfigure.condition.OnWebApplicationCondition.isServletWebApplication(OnWebApplicationCondition.java:100)
	at org.springframework.boot.autoconfigure.condition.OnWebApplicationCondition.isWebApplication(OnWebApplicationCondition.java:75)
	at org.springframework.boot.autoconfigure.condition.OnWebApplicationCondition.getMatchOutcome(OnWebApplicationCondition.java:53)
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:47)
	... 62 more
```

We are using:
Spring Boot 2.0.2.RELEASE
Spring Framework 5.0.6.RELEASE
Spring Security 5.0.5.RELEASE

Everything works fine with SpringFox 2.8.0.  It is only when we upgrade to 2.9.0 that we encounter problems.

**Comment by james-bryant** [comment]
Ok, this seems to only be happening with our test cases.  It seems we can get our Spring Boot app to start up.  It is just when it tries to start it as part of testing.

**Comment by brendanhatton** [comment]
Looks like this (at least in part) caused by a regression in SpringBoot - some more details are here: https://github.com/spring-projects/spring-boot/issues/13226

The ```@EnableSwagger2``` annotation has added a ```@ConditionalOnWebApplication``` annotation in 2.9, which exposes this bug.

**Comment by dilipkrish** [comment]
@brendanhatton thanks for researching that! Saved me a good bit of time!

**Comment by wilkinsona** [comment]
The problem in Spring Boot aside, I don't think it's a good idea for an `@Enable…` annotation to be `@ConditionalOn…`.

I suspect that the intent is to make enabling SpringFox conditional on the application being a web application. What's actually happening is that the class which has been annotated with `@EnableSwagger2` is conditional on the application being a web application. In other words, you are switching off both the user's configuration class and SpringFox when the application is not a web application. I'd recommend making `Swagger2DocumentationConfiguration` `@ConditionalOnWebApplication` instead.

**Comment by dilipkrish** [comment]
Thanks @wilkinsona. That makes sense, its a little subtle and wouldn't have realized it; for that matter, even caught it without your input! Thanks for clarifying. 

A bit of background ... the main reason that we added the conditional was because the `EnableSwagger2` annotation which pulls in the springfox configuration, was causing problems in integration tests that did not need/have a web environment.

I think I understand your comment on moving it to `Swagger2DocumentationConfiguration `, but just to clarify how `@Conditiona..` works; as it is currently setup, springfox will always be enabled only on web applications, so coincidentally this "misconfiguration" happens to work since it is always used in the context of a web application. Is that correct?

**Comment by james-bryant** [comment]
Thanks for the suggestion!  What I did was move the @EnableSwagger2 to the
Swagger2DocumentationConfiguration class and added
@ConditionalOnWebApplication to that class too as suggested.  And then the
tests started working.  Thank you very much.  I now understand the problem
much better.

**Comment by dilipkrish** [comment]
@james-bryant would you be up for submitting a PR?

**Comment by james-bryant** [comment]
Actually, I didn't change SpringFox code.  I just changed my code.  I have
a class named SwaggerDocumentationConfig actually. That's where I put the
Docket bean.  I had the @EnableSwagger2 on my main Spring Boot application
class.  I moved it to this other configuration class and added the
@ConditionalOnWebApplication.  Now when I execute test cases it properly
loads my main class and all the other config classes but ignores the
SwaggerDocumentationConfig class and therefore the @EnableSwagger2.
But, if you want, I can certainly look into fixing this in the actual code
and submitting a PR.  I just won't be able to do this right away.  We're at
the end of a release cycle.
